sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_sender.py

```diff
@@ -349,25 +359,39 @@ def send_messages(self, message, **kwargs):
         timeout = kwargs.pop("timeout", None)
         if timeout is not None and timeout <= 0:
             raise ValueError("The timeout must be greater than 0.")
-        message = transform_messages_to_sendable_if_needed(message)
-        try:
-            batch = self.create_batch()
-            batch._from_list(message)  # pylint: disable=protected-access
-            message = batch
-        except TypeError:  # Message was not a list or generator.
-            pass
-        if isinstance(message, BatchMessage) and len(message) == 0:  # pylint: disable=len-as-condition
-            raise ValueError("A BatchMessage or list of Message must have at least one Message")
-        if not isinstance(message, BatchMessage) and not isinstance(message, Message):
-            raise TypeError("Can only send azure.servicebus.<BatchMessage,Message> or lists of Messages.")
-
-        self._do_retryable_operation(
-            self._send,
-            message=message,
-            timeout=timeout,
-            operation_requires_timeout=True,
-            require_last_exception=True
-        )
+
+        with send_trace_context_manager() as send_span:
+            # Ensure message is sendable (not a ReceivedMessage), and if needed (a list) is batched. Adds tracing.
+            message = transform_messages_to_sendable_if_needed(message)
+            try:
+                for each_message in iter(message): # type: ignore # Ignore type (and below) as it will except if wrong.
+                    add_link_to_send(each_message, send_span)
+                batch = self.create_batch()
+                batch._from_list(message, send_span) # type: ignore # pylint: disable=protected-access
+                message = batch
+            except TypeError:  # Message was not a list or generator. Do needed tracing.
+                if isinstance(message, BatchMessage):
+                    for batch_message in message.message._body_gen:  # pylint: disable=protected-access
+                        add_link_to_send(batch_message, send_span)
```

Comment:
do we need to call `trace_message(message, send_span)` when users passes a BatchMessage object?
cuz if users add message into the batch object by themselves, I think the `add` impl is like `trace_message(message, parent_span=None`.

Reply:
It isn't in EventHubs at least, and I can see a degree of logic to it in that the message was already traced in potentially a different span when added to the batch; vs. the "implicit create batch" logic where we do trace it.  We can check with the other languages.

Reply:
I kinda understand "the message was already traced in potentially a different span"..
but still a bit weird for me to understand as sending is happening within the big `send_span` while separate messages owning their own different span.

basically I'm coming from "how would the trace data be like in the monitor service, do users analyze trace data by "span"?"
well, you're the expert at tracing, maybe my concern doesn't make sense at all.

Reply:
So I double checked JS and it appears they do the same (sender.ts:213).  My understanding is that the "message span" is more in the context of "this message was consumed"; whether by adding it to a batch, or by sending it verbatim, that's when the Message was "first activated."  Candidly while I can "justify" the above to myself I'm not familiar enough as a user to be fully confident, so I'll touch base with Richard to see if he has my same interpretation.

Reply:
Pinging @chradek on this - he worked on this for EventHubs (which is where our "technique" came from) and might be able to provide some more background.

Reply:
> do we need to call trace_message(message, send_span) when users passes a BatchMessage object?
cuz if users add message into the batch object by themselves, I think the add impl is like trace_message(message, parent_span=None.

In JS, we are doing things a bit differently, we create the sendSpan using the message span contexts to link. Python seems to be using the send_span as a parent for the messages. 

As mentioned in the other comment https://github.com/Azure/azure-sdk-for-python/pull/14777#discussion_r521734193, guidelines say `SDK MUST set links before starting the send span.`

@chradek

Reply:
Confirmed with Chris that we're not doing anything super outlandish, but I don't want to break him into jail by saying he was an infallable human trace processor, just gave me enough confidence that this should be safe to GA with.  (Thanks again Chris!)